core/type/base_param_struct.ex

```diff
@@ -0,0 +1,386 @@
+# Copyright(c) 2015-2024 ACCESS CO., LTD. All rights reserved.
+
+use Croma
+
+defmodule AntikytheraCore.BaseParamStruct do
+  @moduledoc """
+  Base module to define a struct that represents abstract parameters.
+  """
+
+  alias Croma.Result, as: R
+
+  @type json_value_t ::
+          boolean
+          | number
+          | String.t()
+          | [json_value_t]
+          | %{String.t() => json_value_t}
+
+  @type params_t :: %{(atom | String.t()) => json_value_t}
+  @type validate_error_reason_t :: :invalid_value | :value_missing
+  @type validate_error_t :: {validate_error_reason_t, [module | {module, atom}]}
+
+  @typep validatable_t :: term
+  @typep preprocessor_t :: (nil | json_value_t -> R.t(validatable_t) | validatable_t)
+  @typep accept_case_t :: :snake | :lower_camel | :upper_camel | :capital
+  @typep default_value_opt_t :: R.t(term, :no_default_value)
+  @typep field_option_t :: {:default, term}
+  @typep mod_with_options_t ::
+           module
+           | {module, preprocessor_t}
+           | {module, [field_option_t]}
+           | {module, preprocessor_t, [field_option_t]}
+  @typep field_t :: {atom, mod_with_options_t}
+  @typep field_with_attr_t ::
+           {atom, [atom], module, preprocessor_t, default_value_opt_t}
+
+  @doc false
+  defun attach_attributes_to_field(
+          {field_name, mod_with_options} :: field_t,
+          accept_case :: nil | accept_case_t,
+          pp_generator :: (module -> R.t(term, :no_default_preprocessor))
+        ) :: field_with_attr_t do
+    accepted_field_names = compute_accepted_field_names(field_name, accept_case)
+
+    {mod, preprocessor, default_value_opt} =
+      extract_mod_and_options(mod_with_options, pp_generator)
+
+    {field_name, accepted_field_names, mod, preprocessor, default_value_opt}
+  end
+
+  defunp extract_mod_and_options(
+           mod_with_options :: mod_with_options_t,
+           pp_generator :: (module -> R.t(term, :no_default_preprocessor))
+         ) :: {module, preprocessor_t, default_value_opt_t} do
+    {mod, preprocessor, [default: default_value]}, _pp_generator
+    when is_atom(mod) and is_function(preprocessor, 1) ->
+      {mod, preprocessor, {:ok, default_value}}
+
+    {mod, preprocessor}, _pp_generator when is_atom(mod) and is_function(preprocessor, 1) ->
+      {mod, preprocessor, compute_default_value(mod)}
+
+    {mod, [default: default_value]}, pp_generator when is_atom(mod) ->
+      {mod, R.get(pp_generator.(mod), &Function.identity/1), {:ok, default_value}}
+
+    mod, pp_generator when is_atom(mod) ->
+      {mod, R.get(pp_generator.(mod), &Function.identity/1), compute_default_value(mod)}
+  end
+
+  defunp compute_default_value(mod :: v[module]) :: default_value_opt_t do
```

Comment:
`compute` is ambiguous. This function gets a default value from `mod`. I think `(try_)get_default_value_from_mod` or `_module` is better.

Reply:
Fixed at f69e267.